lib/user_communication_defender.rb

```diff
@@ -0,0 +1,123 @@
+    # disabled PMs from them or anyone at all.
+    user_communication_preferences.each do |user|
+      resolved_user_communication_preferences[user.id] = UserCommunicationPreference.new(
+        username: @target_users[user.id],
+        is_muting: user.is_muting,
+        is_ignoring: user.is_ignoring,
+        is_disallowing_all_pms: user.is_disallowing_all_pms,
+        is_disallowing_pms_from_acting_user: false
+      )
+    end
+
+    # If any of the users has allowed_pm_users enabled check to see if the creator
+    # is in their list.
+    users_with_allowed_pms = user_communication_preferences.select(&:enable_allowed_pm_users)
+    if users_with_allowed_pms.any?
+
+      user_ids_with_allowed_pms = users_with_allowed_pms.map(&:id)
+      user_ids_acting_can_pm = AllowedPmUser.where(
+        allowed_pm_user_id: @acting_user.id, user_id: user_ids_with_allowed_pms
+      ).pluck(:user_id).uniq
+
+      # If not in the list mark them as not accepting communication.
+      user_ids_acting_cannot_pm = user_ids_with_allowed_pms - user_ids_acting_can_pm
+      user_ids_acting_cannot_pm.each do |user_id|
+        if resolved_user_communication_preferences[user_id]
+          resolved_user_communication_preferences[user_id].is_disallowing_pms_from_acting_user = true
+        else
+          resolved_user_communication_preferences[user_id] = UserCommunicationPreference.new(
+            username: @target_users[user_id],
+            is_muting: false,
+            is_ignoring: false,
+            is_disallowing_all_pms: false,
+            is_disallowing_pms_from_acting_user: true
+          )
+        end
+      end
+    end
+
+    UserCommunicationPreferences.new(@acting_user, resolved_user_communication_preferences)
+  end
+
+  private
+
+  def user_communication_preferences
+    @user_communication_preferences ||= DB.query(<<~SQL, acting_user_id: @acting_user.id, target_user_ids: @target_users.keys)
+      SELECT users.id,
+      CASE WHEN muted_users.muted_user_id IS NOT NULL THEN true ELSE false END AS is_muting,
+      CASE WHEN ignored_users.ignored_user_id IS NOT NULL THEN true ELSE false END AS is_ignoring,
+      CASE WHEN user_options.allow_private_messages THEN false ELSE true END AS is_disallowing_all_pms,
+      user_options.enable_allowed_pm_users
+      FROM users
+      LEFT JOIN user_options ON user_options.user_id = users.id
+      LEFT JOIN muted_users ON muted_users.user_id = users.id AND muted_users.muted_user_id = :acting_user_id
+      LEFT JOIN ignored_users ON ignored_users.user_id = users.id AND ignored_users.ignored_user_id = :acting_user_id
+      WHERE (user_options.user_id IS NOT NULL AND user_options.user_id IN (:target_user_ids)) AND
+      (
+        NOT user_options.allow_private_messages OR
+        user_options.enable_allowed_pm_users OR
+        muted_users.user_id IN (:target_user_ids) OR
+        ignored_users.user_id IN (:target_user_ids)
+      )
```

Comment:
```suggestion
      LEFT JOIN user_options ON user_options.user_id = users.id
      LEFT JOIN muted_users ON muted_users.user_id = users.id AND muted_users.muted_user_id = :acting_user_id
      LEFT JOIN ignored_users ON ignored_users.user_id = users.id AND ignored_users.ignored_user_id = :acting_user_id
      WHERE users.id IN (:target_user_ids) AND
      (
        NOT user_options.allow_private_messages OR
        user_options.enable_allowed_pm_users OR
        muted_users.user_id IS NOT NULL OR
        ignored_users.user_id IS NOT NULL
      )
```

I think you might be able to simplify the query here to the above.

Reply:
I will give it a try but we need to make sure `user_options` is there...maybe I can `INNER JOIN` user options

Reply:
This worked :+1: